server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -319,6 +359,111 @@ final long bucketOrd(long owningBucketOrdinal, int filterOrd) {
         }
     }
 
+    /**
+     * The competitive iterator wraps the per-filter iterators, skipping rangers with no filter matches during iteration.
+     * This is achieved by storing the doc matches that each iterator returns in a priority queue that's
+     * then used to provide the next possible matches during advancing.
+     */
+    private static class CompetitiveIterator extends DocIdSetIterator {
+
+        final DocIdSetIterator[] filteredDocIdSetIterators;
+        final int[] lastMatches;
+        final DocId currentDocID;
+        final int maxDocId;
+
+        // Tracks next doc matches as they get returned from the per-filter iterators.
+        final PriorityQueue<Integer> nextDocs;
```

Comment:
Maybe use Lucene's `PriorityQueue` instead here and that contains `DocIdSetIterator` instances?
(`PriorityQueue#lessThan()` would compare `DocIdSetIterartor#docId()` and each time a `DocIdSetIterartor #advance(...)` is invoked `PriorityQueue#updateTop(...)` should be invoked )

Reply:
(this pattern is used in other places as well)

Reply:
Hm that may be somewhat tricky to use as we want to keep the last doc per filter to update the corresponding bucket. Let's discuss this offline in our sync.

Reply:
I think we can extract the last docid from this priority queue and use it in several places. But let's chat more about this in the sync.